src/main/scala/com/microsoft/hyperspace/actions/DeleteOnReadAction.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.telemetry.{AppInfo, DeleteOnReadActionEvent, HyperspaceEvent}
+
+/**
+ * Refresh index by updating list of excluded source data files and index signature
+ * in index metadata.
+ * Note this Refresh Action only fixes an index metadata w.r.t deleted source data files
```

Comment:
I also see log messages like "Refresh aborted", "Refresh index is updating index metadata", "during index refresh", etc.

Reply:
After going thru the code once, it seems that this should be "RefreshSomethingAction" than "DeleteOnReadAction".

Reply:
I changed it to `RefreshForDeleteOnReadAction` to capture the fact that it is a Refresh action.

Reply:
Hmm, should we just call this `RefreshLogEntryAction` and update both deleted and appended files? (I think appended can be leveraged during read with @sezruby's recent PR?)

To me, having an transaction log action that has a dependency on the "read" behavior does not sound right.

Reply:
@imback82 Do you mean beside `excluded`, in this PR we also add a new property to `IndexLogEntry` as `appended` and populate it?

Reply:
Yes, both `excluded` and `appended` can be utilized right?

Reply:
Yes, `excluded` is utilized for delete - `appended` "could" be utilized for HybridScan, specifically for below code.
This should work as signature always makes sure that we pick an index if it can replace a table scan done on the current/latest set of files. (@sezruby can comment on this).

```
val curFileSet = location.allFiles
          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
val filesAppended =
          curFileSet.filterNot(index.allSourceFileInfos.contains).map(f => new Path(f.name))
```

Reply:
I renamed `RefreshForDeleteOnReadAction` -> `RefreshLogEntryAction` and added `appended` property to IndexLogEntry.
However, we should now rename `RefreshDeleteActionBase` to sth else as one of its children: `RefreshLogEntryAction` is no longer just addressing deleted files. I could not come-up with a better name for `RefreshDeleteActionBase`, any suggestions?

Reply:
We could utilize the current code base - check if the index entry has `appended` or `excluded` files and use them instead filesAppended or filesDeleted in `transformPlanToUseHybridScan`(https://github.com/microsoft/hyperspace/pull/171/files#diff-1ac09a3b18348fa9bf89fd4c42f08390R237) - then #175 might be unnecessary?
To do this, we need to make sure that there is no diff in the source file list at query time. 

BTW I think, when hybrid scan config is enabled: 
1.  check `signatureValid` first in `getCandidateIndex`
2. if there's no candidate, try to get candidate index with `isHybridScanCandidate`

But I'm not sure this could be best - for example, it's might be more efficient to avoid full-shuffling of index data. So an index pair **with proper bucketing and hybrid scan** _might be more efficient_ than full index pair **without proper bucketing (no hybrid scan)**. I guess rank algorithm for hybrid scan might be more complex than I thought in #164 ..

Reply:
> We could utilize the current code base - check if the index entry has `appended` or `excluded` files and use them instead filesAppended or filesDeleted in `transformPlanToUseHybridScan`(https://github.com/microsoft/hyperspace/pull/171/files#diff-1ac09a3b18348fa9bf89fd4c42f08390R237)

Yes, this was exactly what I was thinking.

> To do this, we need to make sure that there is no diff in the source file list at query time.

I think we can utilize the signature this PR computes; it included new files as well.

BTW, should we call these fields as `added` (maybe `appended`) and `deleted` instead? WDYT? cc: @rapoth

Reply:
+1 to both code reuse and appended/deleted.

Reply:
@pirz @imback82 @sezruby I thought more about this. It would be good to not have duplicate code paths for delete. For instance, hybrid scan is already taking care of applying the delete predicates appropriately so we should not have a secondary path do this. Same thing with appends (if applicable). This will save us from future maintenance hell :)

CC: @apoorvedave1

Reply:
@rapoth "support for delete on read" and "support for hybrid scan" are triggered via separate configs (at least for now). I assume if we combine them then one can only benefit from "delete" on read if "hybrid scan" is enabled for the index and they may bring some other stuff. I believe we originally had a discussion about it (not adding "delete on read" and relying on Hybridscan, but the decision was these are different actions and they are not necessarily targeting same customers).
I prefer to keep them separate and let the PRs move forward. If we wanna combine them, we can do that once we drop (temp) feature flags and finalize the API calls.  @rapoth  What do you think?